ipam/gossip.go

```diff
@@ -0,0 +1,78 @@
+package ipam
+
+import (
+	"github.com/zettio/weave/router"
+)
+
+// GossipData implementation is trivial - we always gossip the whole ring
+type ipamGossipData struct {
+	alloc *Allocator
+}
+
+func (d *ipamGossipData) Merge(other router.GossipData) {
+	// no-op
+}
+
+func (d *ipamGossipData) Encode() []byte {
+	return d.alloc.Encode()
+}
+
+// Gossip returns a GossipData implementation, which in this case always
+// returns the latest ring state (and does nothing on merge)
+func (alloc *Allocator) Gossip() router.GossipData {
+	return &ipamGossipData{alloc}
+}
+
+// SetGossip gives the allocator an interface for talking to the outside world
+func (alloc *Allocator) SetGossip(gossip router.Gossip) {
+	alloc.gossip = gossip
+}
+
+// OnGossipUnicast (Sync)
+func (alloc *Allocator) OnGossipUnicast(sender router.PeerName, msg []byte) error {
+	alloc.debugln("OnGossipUnicast from", sender, ": ", len(msg), "bytes")
+	resultChan := make(chan error)
+	alloc.actionChan <- func() {
+		switch msg[0] {
+		case msgLeaderElected:
+			resultChan <- alloc.handleLeaderElected()
+		case msgSpaceRequest:
+			// some other peer asked us for space
+			alloc.donateSpace(sender)
+			resultChan <- nil
+		case msgRingUpdate:
+			resultChan <- alloc.updateRing(msg[1:])
+		}
+	}
+	return <-resultChan
+}
+
+// OnGossipBroadcast (Sync)
+func (alloc *Allocator) OnGossipBroadcast(msg []byte) (router.GossipData, error) {
+	alloc.debugln("OnGossipBroadcast:", len(msg), "bytes")
+	resultChan := make(chan error)
+	alloc.actionChan <- func() {
+		resultChan <- alloc.updateRing(msg)
+	}
+	return alloc.Gossip(), <-resultChan
+}
+
+// Encode (Sync)
+func (alloc *Allocator) Encode() []byte {
+	resultChan := make(chan []byte)
+	alloc.actionChan <- func() {
+		resultChan <- alloc.ring.GossipState()
+	}
+	return <-resultChan
+}
+
+// OnGossip (Sync)
+func (alloc *Allocator) OnGossip(msg []byte) (router.GossipData, error) {
+	alloc.debugln("Allocator.OnGossip:", len(msg), "bytes")
+	resultChan := make(chan error)
+	alloc.actionChan <- func() {
+		resultChan <- alloc.updateRing(msg)
+	}
+	err := <-resultChan
```

Comment:
sorry, can you re-explain "inline this receive below"; it may have lost context.